substratevm/DebugInfo.md

```diff
@@ -90,12 +165,346 @@ The currently implemented features include:
+fields from an array header struct specific to the array element type,
+one of _arrhdrZ _arrhdrB, _arrhdrS, ... _arrhdrA (the last one is for
+object arrays). Inherited fields include the hub, array length, idHash
+to round up the header size to a boundary suitable for the array
+element type. The array class (layout) type includes only one field, a
+C++ array of length zero whose element type is a primtiive type or
+Java referece type.
+
+```
+(gdb) ptype 'java.lang.String[]'
+type = struct java.lang.String[] : public _arrhdrA {
+    java.lang.String *data[0];
+}
+```
+
+Notice that the type of the values stored in the data array is
+`java.lang.String *` i.e. the C++ array stores Java references
+i.e. addresss as far as the C++ model is concerned.
+
+The array header structs are all extensions of the basic _objhdr type
+which means that arrays and objects can both be safely cast to oops.
+
+```
+(gdb) ptype _arrhdrA
+type = struct _arrhdrA {
+    java.lang.Class *hub;
+    int len;
+    int idHash;
+}
+```
+
+Interfaces layouts are modelled as union types whose members are the
+layouts for all the classes which implement the interfacse.
+
+```
+(gdb) ptype 'java.lang.CharSequence'
+type = union _java.lang.CharSequence {
+    _java.lang.AbstractStringBuilder _java.lang.AbstractStringBuilder;
+    _java.lang.StringBuffer _java.lang.StringBuffer;
+    _java.lang.StringBuilder _java.lang.StringBuilder;
+    _java.lang.String _java.lang.String;
+    _java.nio.CharBuffer _java.nio.CharBuffer;
+}
+```
+
+Given a reference typed to an interface it can be resolved to the
+relevant class type by viewing it through the relevant union element.
+
+If we take the first String in the args array we can ask gdb to cast
+it to interface CharSequence
+```
+(gdb) print (('java.lang.String[]' *)$rdi)->data[0]
+$6 = (java.lang.String) 0x7ffff7c01060
+(gdb) print ('java.lang.CharSequence')$6
+$7 = (java.lang.CharSequence) 0x7ffff7c01060
+```
+
+The hubname command can be used to identify the actual type of the
+object that implements this interface and that type name can be used
+to select the union element used to print the object.
+
+```
+(gdb) hubname $7
+0x7d96d8:	"java.lang.String\270", <incomplete sequence \344\220>
+(gdb) print $7->'_java.lang.String'
+$18 = {
+  <java.lang.Object> = {
+    <_objhdr> = {
+      hub = 0x90cb58
+    }, <No data fields>}, 
+  members of java.lang.String:
+  value = 0x7ffff7c011a0,
+  hash = 0,
+  coder = 0 '\000'
+}
+```
+
+The current debug info model does not include the location info needed
+to allow symbolic names for local vars and parameter vars to be
+resolved to primitive values or object references. However, the
+debugger does understand method names and static field names.
+
+The following command places a breakpoint on the main entry point for
+class `Hello`. Note that since GDB thinks this is a C++ method it uses
+the `::` separator to separate the method name from the class name.
```

Comment:
Nice advantage of using `::` as separator. Makes it dead simple to grep for method names.

Reply:
Yes, Java probably has too many dots for its own good :-) Other than that the syntax is perf... oh wait!